packages/compass-query-bar/src/components/option-editor/option-editor.tsx

```diff
@@ -0,0 +1,183 @@
+import React, { Component } from 'react';
+import { QueryAutoCompleter } from 'mongodb-ace-autocompleter';
+import {
+  Editor,
+  EditorVariant,
+  EditorTextCompleter,
+  css,
+  cx,
+  focusRingStyles,
+  focusRingVisibleStyles,
+  uiColors,
+  spacing,
+} from '@mongodb-js/compass-components';
+import type { Listenable } from 'reflux';
+
+import type { QueryOption as QueryOptionType } from '../../constants/query-option-definition';
+
+const editorStyles = cx(
+  css({
+    '&::before': {
+      position: 'absolute',
+      content: '""',
+      pointerEvents: 'none',
+      top: 3,
+      right: 3,
+      bottom: 3,
+      left: 3,
+      borderRadius: spacing[1],
+      boxShadow: 'inset 0px 0px 0px 0px transparent',
+      transition: 'box-shadow .16s ease-in',
+    },
+    borderRadius: '6px',
+    '&:hover': {
+      '&::after': {
+        boxShadow: `0 0 0 3px ${uiColors.gray.light2}`,
+        transitionTimingFunction: 'ease-out',
+      },
+    },
+    '&:focus-within': focusRingVisibleStyles,
+  }),
+  focusRingStyles
+);
+
+const editorWithErrorStyles = css({
+  '&:focus-within': {
+    '&::before': {
+      boxShadow: 'inset 0 0 0 1px transparent',
+    },
+  },
+  '&::before': {
+    zIndex: 5,
+    boxShadow: `inset 0px 0px 0px 1px ${uiColors.red.base}`,
+  },
+});
+
+type OptionEditorProps = {
+  autoPopulated: boolean;
+  hasError: boolean;
+  id: string;
+  onChange: (value: string) => void;
+  onApply: () => void;
+  placeholder?: string;
+  queryOption: QueryOptionType;
+  refreshEditorAction: Listenable;
+  schemaFields?: string[];
+  serverVersion?: string;
+  value?: any;
+};
+
+export class OptionEditor extends Component<OptionEditorProps> {
+  completer: any;
+  boundOnFieldsChanged: (schemaFields?: string[]) => void;
+  unsub?: () => void;
+  editor: any;
+
+  static defaultProps = {
+    queryOption: '',
+    value: '',
+    serverVersion: '3.6.0',
+    autoPopulated: false,
+    schemaFields: [],
+  };
+
+  /**
+   * Set up the autocompleters once on initialization.
+   *
+   * @param {Object} props - The properties.
+   */
+  constructor(props: OptionEditorProps) {
+    super(props);
+    this.completer = new QueryAutoCompleter(
+      props.serverVersion,
+      EditorTextCompleter,
+      props.schemaFields
+    );
+    this.boundOnFieldsChanged = this.onFieldsChanged.bind(this);
+  }
+
+  /**
+   * Subscribe on mount.
+   */
+  componentDidMount(): void {
+    this.unsub = this.props.refreshEditorAction.listen(() => {
+      this.editor.setValue(this.props.value);
+      this.editor.clearSelection();
+    });
+  }
+
+  /**
+   * @param {Object} nextProps - The next properties.
+   *
+   * @returns {Boolean} If the component should update.
+   */
+  shouldComponentUpdate(nextProps: OptionEditorProps): boolean {
+    this.boundOnFieldsChanged(nextProps.schemaFields);
```

Comment:
`shouldComponentUpdate` is really not a great place for side-effects. What's exactly is happening here? Why are we always updating completer but not allowing component to update?

Reply:
Thanks for bringing it up, I had copied this over without going too deep into why we were doing it this way.
From my understanding it's a bit of a performance boost mixed with a workaround for non-displayed ace editor behavior. I think we can do better here. What previously would happen is the editor would be initialized and then its value would only be controlled when the `refreshEditor` action occurred. This action happened when the `subtab-changed` global app registry event occurs here: https://github.com/mongodb-js/compass/blob/d78f63c28ed5ff516ce6453c44245f5e020c7b70/packages/compass-query-bar/src/stores/query-bar-store.js#L760
That's emitted when a user is changing from the Schema tab to the Documents tab for instance. When that happened we would then set the editor's value to the new value: https://github.com/mongodb-js/compass/blob/d78f63c28ed5ff516ce6453c44245f5e020c7b70/packages/compass-query-bar/src/components/option-editor/option-editor.jsx#L54
We were doing this because when the ace editor is not visible (not the active subtab in Compass), [ace does not update the view](https://github.com/ajaxorg/ace/issues/2497). So we would perform this `value` setting after the subtab was changed to ensure we showed the new value after the editor's container has become visible (I'm thinking this could lead to race conditions).

I'm investigating how we can ensure the view is updated when the view is visible without unrelated global event chain listening. It may be a bit out of the scope of this pr (in which case would it be alright to keep the existing implementation? With added comments)

Reply:
Looks like we won't need the `autoPopulated` flag if we do this differently also.

Reply:
Pushed an update with a comment on the behavior and why we do it. Cleaned up this component a bit as well. I think other solutions which would allow us to remove the global event listening would be out of the scope of this work, interested to see how you feel on it though.